bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/internal/DeviceManagerFactory.java

```diff
@@ -0,0 +1,174 @@
+ * In addition, this class is responsible for managing the BlueZPropertiesChangedHandler instance
+ * used by the binding for listening and dispatching dbus events from the DeviceManager.
+ *
+ * Creation of the DeviceManager is asynchronous and thus attempts to retrieve the DeviceManager
+ * through 'getDeviceManager' may initially fail.
+ *
+ * @author Connor Petty - Initial Contribution
+ *
+ */
+@NonNullByDefault
+@Component(service = DeviceManagerFactory.class)
+public class DeviceManagerFactory {
+
+    private final Logger logger = LoggerFactory.getLogger(DeviceManagerFactory.class);
+    private final BlueZPropertiesChangedHandler changeHandler = new BlueZPropertiesChangedHandler();
+
+    private @Nullable ScheduledExecutorService scheduler;
+    private @Nullable DeviceManager deviceManager;
+    private @Nullable CompletableFuture<DeviceManagerWrapper> deviceManagerFuture;
+
+    public BlueZPropertiesChangedHandler getPropertiesChangedHandler() {
+        return changeHandler;
+    }
+
+    public @Nullable DeviceManagerWrapper getDeviceManager() {
+        // we can cheat the null checker with casting here
+        var future = (CompletableFuture<@Nullable DeviceManagerWrapper>) this.deviceManagerFuture;
+        if (future != null) {
+            return future.getNow(null);
+        }
+        return null;
+    }
+
+    @Activate
+    public void initialize() {
+        logger.debug("initializing DeviceManagerFactory");
+        ScheduledThreadPoolExecutor scheduler = new ScheduledThreadPoolExecutor(1,
+                new NamedThreadFactory("bluetooth.bluez-init", true));
+
+        scheduler.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        scheduler.setRemoveOnCancelPolicy(true);
+
+        this.scheduler = scheduler;
+
+        this.deviceManagerFuture = callAsync(() -> {
+            try {
+                // if this is the first call to the library, this call
+                // should throw an exception (that we are catching)
+                return DeviceManager.getInstance();
+                // Experimental - seems reuse does not work
+            } catch (IllegalStateException e) {
+                // Exception caused by first call to the library
+                return DeviceManager.createInstance(false);
+            }
+        }, scheduler)//
+                .thenApply(devManager -> {
+                    this.deviceManager = devManager;
+                    return devManager;
+                }).thenCompose(devManager -> registerPropertyHandler(devManager, scheduler))//
+                .thenApply(DeviceManagerWrapper::new)//
+                .whenComplete((devManager, th) -> {
+                    if (th != null) {
+                        logger.warn("Failed to initialize DeviceManager: {}", th.getMessage());
+                    }
+                });
+    }
+
+    @Deactivate
+    public void dispose() {
+        DeviceManager manager = deviceManager;
```

Comment:
> Is there a concurrency issue that forces you to assign field to local variable?

Bingo, we use a pretty strict NPE prevention tool that will ignore null checks for `@Nullablee` fields.

Reply:
Are you kidding me? Does the tool is blind to non null condition or imply multi thread access based on code in other places?

Anyhow - maybe it is avoidable with `AtomicReference.getAndUpdate` which should be consider safer?
Looking at code and use of `CompletableFuture` you might be still better with direct access over future which should retain computed value unless disposed..

Reply:
> Are you kidding me? Does the tool is blind to non null condition or imply multi thread access based on code in other places?

I think the tool it taking the heavy-handed assumption that any nullable field could be changed to null by another thread at any time. Overzealous and inappropriate in a vast majority of situations but if you hate NPE enough the tradeoff can be worth it.

> Anyhow - maybe it is avoidable with AtomicReference.getAndUpdate which should be consider safer?
Looking at code and use of CompletableFuture you might be still better with direct access over future which should retain computed value unless disposed..

My use of completable future was to prevent access to the `deviceManager` before the BlueZPropertiesChangedHandler was registered. Making sure that BlueZPropertiesChangedHandler successfully registered is a rather good litmus test for whether the deviceManager is in a state usable by the rest of the application.

I suppose I could add more fields to keep track of registration state but I think adding more state fields might make the code less maintainable.